File: tests/ui-tests/src/factories/user.factory.ts
import { UserBuilder } from "../builders/user.builder";
import type { TestUser } from "../models/TestUser";
import type { UserCreationStrategy } from "../strategies/user/user-creation.strategy";

export class UserFactory {
    async create(
        strategy: UserCreationStrategy,
        configure?: (b: UserBuilder) => UserBuilder
    ): Promise<TestUser> {
        const builder = new UserBuilder();
        const configured = configure ? configure(builder) : builder;
        return strategy.createUser(configured.build());
    }

    async createMinimal(
        strategy: UserCreationStrategy,
        configure?: (b: UserBuilder) => UserBuilder
    ): Promise<TestUser> {
        let builder: UserBuilder = new UserBuilder().required();
        if (configure) {
            builder = configure(builder);
        }
        return strategy.createUser(builder.build());
    }
}

File: tests/backend_tests/src/factories/user_factory.py
"""
Factory that composes UserBuilder with a UserCreationStrategy (mirrors tests/ui-tests/src/factories/user.factory.ts).
"""

from __future__ import annotations

from typing import Callable, Optional

from builders.user_builder import UserBuilder
from strategies.user.user_creation_strategy import TRegisteredUser, UserCreationStrategy


class UserFactory:
    """Strategy-first API (same argument order as ``tests/ui-tests/src/factories/user.factory.ts``)."""

    def create(
        self,
        strategy: UserCreationStrategy[TRegisteredUser],
        configure: Optional[Callable[[UserBuilder], UserBuilder]] = None,
    ) -> TRegisteredUser:
        builder = UserBuilder()
        if configure:
            builder = configure(builder)
        return strategy.create_user(builder.build())

    def create_minimal(
        self,
        strategy: UserCreationStrategy[TRegisteredUser],
        configure: Optional[Callable[[UserBuilder], UserBuilder]] = None,
    ) -> TRegisteredUser:
        builder: UserBuilder = UserBuilder().required()
        if configure:
            builder = configure(builder)
        return strategy.create_user(builder.build())
